### PR TITLE
Add URL expander plugin

### PR DESCRIPTION
**Pull request by Ackar**

### CODE REVIEW

File: lib/plugins/urlExpander.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+var log = require('../log');
+
+var NAME = exports.name = 'urlExpander';
+
+var metrics = require('../metrics').session(NAME);
+
+var hosts = {};
+
+exports.init = function() {
+  fs.readFile('config/urlShorteners.txt', function(err, data) {
```
Comment (eamsen) [inline lib/plugins/urlExpander.js]: Better make that configurable.

File: test/tests/urlexpander.js [hunk 2]
```diff
@@ -0,0 +1,87 @@
+'use strict';
```
Comment (eamsen) [inline test/tests/urlexpander.js:1]: We have a mix of camel case and just lower case file names, please stick to one and be consistent.
Might be worth to go through the repository and correct that via `git mv`.

File: config/default.yml [hunk 3]
```diff
@@ -124,6 +119,12 @@ adblock:
   optional: true
   listUrl: 'http://pgl.yoyo.org/adservers/serverlist.php?&mimetype=plaintext'
 
+urlExpander:
```
Comment (eamsen) [inline config/default.yml]: Please add a short comment describing what it does.

File: lib/plugins/urlexpander.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+var NAME = exports.name = 'urlExpander';
+
+var metrics = require('../metrics').session(NAME);
+
+var hosts = {};
+
+exports.init = function() {
+  fs.readFile(config.urlExpander.hostsFile, function(err, data) {
+    if (err) {
+      log.error('Unable to read URL shorteners list');
```
Comment (eamsen) [inline lib/plugins/urlexpander.js]: Add the path in the error message.

File: lib/plugins/urlexpander.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+  var currentUrl = url.parse(request.url);
+
+  if (redirectCount < config.urlExpander.maxRedirect &&
+      currentUrl.hostname in hosts) {
+    http.get(request.url, function(res) {
+      // We need to consume the data so the socket is not kept open.
+      res.on('data', function() {});
+
+      if (res.statusCode >= 300 && res.statusCode < 400) {
+        request.url = res.headers.location;
```
Comment (eamsen) [inline lib/plugins/urlexpander.js:53]: We should probably forward some of the request header fields (e.g. `referer`).

Comment (Ackar) [inline lib/plugins/urlexpander.js:53]: The header fields of the original request are not modified, only the request.url is modified. So if a `referer` field is present it should still be there after that.

Comment (snorp) [inline lib/plugins/urlexpander.js:53]: I think he means in the http.get here. Right now you're not passing any headers at all (except maybe some implicit ones).

**Comment by Ackar** [comment]
Do we want this or not?
To summarize (because the plugin behavior has changed since the pull request was created), if a request to a url shortener domain is received the plugin will resolve the "real" url and send the last redirect response to the client. So the client will receive exactly one 3XX response, pointing to the target url.

I think we may want to have that, but disabled by default since the only use case I can see is for twitter users. For that use case it can be a huge speed improvement though, because a lot of twitter links have at least 3 or 4 redirections.

**Comment by eamsen** [comment]
I don't think we should add plugins, which are only useful for specific sites. But I think this could eventually be made useful in general.

My vote goes for not merging it yet and keeping the branch around for the future, but would not be opposed to merging either. @snorp what do you think?

**Pull request merged**